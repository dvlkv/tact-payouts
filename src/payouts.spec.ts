
import { ContractSystem, Treasure } from '@tact-lang/emulator';
import { Address, beginCell, toNano } from 'ton-core';
import { KeyPair, keyPairFromSeed, sign } from 'ton-crypto';
import { Payouts } from '../output/payouts_Payouts';

jest
    .useFakeTimers()
    .setSystemTime(new Date('2023-01-01'));

function createPayout(receiver: Address, amount: bigint, secretKey: Buffer) {
    let toSign = beginCell().storeAddress(receiver).storeCoins(amount).endCell().hash();
    let signature = sign(toSign, secretKey);
    // Align cell
    let resultSc = beginCell().storeCoins(amount).storeBuffer(signature);
    if (resultSc.bits % 8 != 0) {
        resultSc.storeUint(0, 8 - (resultSc.bits % 8));
    }

    let result = resultSc.endCell();
    return result.bits.subbuffer(0, result.bits.length)!.toString('base64');
}


describe('Payouts', () => {
    let system: ContractSystem;
    let owner: Treasure;
    let keys: KeyPair;

    beforeEach(async () => {
        system = await ContractSystem.create();

        owner = system.treasure('owner');
        system.name(owner, 'owner'); // Set a name for the contract. Useful for snapshot testing.

        keys = keyPairFromSeed(Buffer.from('ac0cebac0cebac0cebac0cebac0cebac'));
    });

    it("should deploy & topup", async () => {
        let startDate = Date.now() + 60 * 60 * 1000;
        let burnTimeout = 60 * 60;
        const contract = system.open(await Payouts.fromInit(owner.address, BigInt('0x' + keys.publicKey.toString('hex')), BigInt(Math.ceil(startDate / 1000)), BigInt(burnTimeout)));

        let track = system.track(contract);

        await contract.send(owner, {
            value: toNano(1),
            bounce: false,
        }, 'Topup');

        await system.run();

        expect(track.collect()).toMatchSnapshot();
    });

    it("should withdraw", async () => {
        let startDate = Date.now() + 60 * 60 * 1000;
        let burnTimeout = 60 * 60;
        const contract = system.open(await Payouts.fromInit(owner.address, BigInt('0x' + keys.publicKey.toString('hex')), BigInt(Math.ceil(startDate / 1000)), BigInt(burnTimeout)));

        await contract.send(owner, {
            value: toNano(100000),
            bounce: false,
        }, 'Topup');

        await system.run();

        let client = system.treasure('client');
        let track = system.track(contract);

        let payload = createPayout(client.address, toNano(10), keys.secretKey);

        await contract.send(client, {
            value: toNano(1),
            bounce: false,
        }, payload);

        await system.run();

        expect(track.collect()).toMatchSnapshot();
    });

    it("should not withdraw twice", async () => {
        let startDate = Date.now() + 60 * 60 * 1000;
        let burnTimeout = 60 * 60;
        const contract = system.open(await Payouts.fromInit(owner.address, BigInt('0x' + keys.publicKey.toString('hex')), BigInt(Math.ceil(startDate / 1000)), BigInt(burnTimeout)));

        await contract.send(owner, {
            value: toNano(100000),
            bounce: false,
        }, 'Topup');

        await system.run();

        let client = system.treasure('client');
        let track = system.track(contract);

        let payload = createPayout(client.address, toNano(10), keys.secretKey);

        await contract.send(client, {
            value: toNano(1),
            bounce: false,
        }, payload);

        await system.run();
        track.collect();

        await contract.send(client, {
            value: toNano(1),
            bounce: false,
        }, payload);

        await system.run();

        expect(track.collect()).toMatchSnapshot();
    });

    it('should start burn', async () => {
        let startDate = Math.ceil((Date.now() + 60 * 60 * 1000) / 1000);
        let burnTimeout = 60 * 60;
        const contract = system.open(await Payouts.fromInit(owner.address, BigInt('0x' + keys.publicKey.toString('hex')), BigInt(startDate), BigInt(burnTimeout)));

        await contract.send(owner, {
            value: toNano(100000),
            bounce: false,
        }, 'Topup');

        await system.run();

        system.update({
            now: startDate + burnTimeout / 2,
        });

        let client = system.treasure('client');
        let track = system.track(contract);

        let payload = createPayout(client.address, toNano(10), keys.secretKey);

        await contract.send(client, {
            value: toNano(1),
            bounce: false,
        }, payload);

        await system.run();
        expect(track.collect()).toMatchSnapshot();
    });

    it('should burn completely', async () => {
        let startDate = Math.ceil((Date.now() + 60 * 60 * 1000) / 1000);
        let burnTimeout = 60 * 60;
        const contract = system.open(await Payouts.fromInit(owner.address, BigInt('0x' + keys.publicKey.toString('hex')), BigInt(startDate), BigInt(burnTimeout)));

        await contract.send(owner, {
            value: toNano(100000),
            bounce: false,
        }, 'Topup');

        await system.run();

        system.update({
            now: startDate + burnTimeout,
        });

        let client = system.treasure('client');
        let track = system.track(contract);

        let payload = createPayout(client.address, toNano(10), keys.secretKey);

        await contract.send(client, {
            value: toNano(1),
            bounce: false,
        }, payload);

        await system.run();
        expect(track.collect()).toMatchSnapshot();
    });
});