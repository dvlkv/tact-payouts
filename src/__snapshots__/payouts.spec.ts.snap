// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payouts should burn completely 1`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "UCVAvkAKQCCn7CYC5oDzJww7/mYFJTKuGwzv7XbfuTQhsZXwBwjwd+ai9d8SUHujsjk/ji2v3XENWFVYrf/K1jLmUmcwgA==",
            "type": "text",
          },
          "bounce": false,
          "from": "@treasure(client)",
          "to": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "failed",
        "errorCode": 20020,
        "errorMessage": "Not enough funds to withdraw",
      },
    ],
  },
]
`;

exports[`Payouts should deploy & topup 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "Topup",
            "type": "text",
          },
          "bounce": false,
          "from": "@owner",
          "to": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 7081n,
      },
    ],
  },
]
`;

exports[`Payouts should not withdraw twice 1`] = `
[
  {
    "$seq": 3,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "UCVAvkAKQCCn7CYC5oDzJww7/mYFJTKuGwzv7XbfuTQhsZXwBwjwd+ai9d8SUHujsjk/ji2v3XENWFVYrf/K1jLmUmcwgA==",
            "type": "text",
          },
          "bounce": false,
          "from": "@treasure(client)",
          "to": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 65561n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{C41949DF000000000000000000000000000000000000000000000000000000012A05F2004_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
            "to": "kQCEU_9CZOdW0nEmPIFBeKm5TwUgQeZSSoGV5F-1NEfHVOro",
            "type": "internal",
            "value": 923332000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 4,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{FFEB40DE000000000000000000000000000000000000000000000000000000012A05F200400A128FE2F818AF4A1202DCB07C4374ED1EF1F35DC49E9FAD6BD48BDFB08E74C404_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCEU_9CZOdW0nEmPIFBeKm5TwUgQeZSSoGV5F-1NEfHVOro",
          "to": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
          "type": "internal",
          "value": 911078000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 17912n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Already paid",
              "type": "text",
            },
            "bounce": false,
            "from": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
            "to": "@treasure(client)",
            "type": "internal",
            "value": 891938000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`Payouts should start burn 1`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "UCVAvkAKQCCn7CYC5oDzJww7/mYFJTKuGwzv7XbfuTQhsZXwBwjwd+ai9d8SUHujsjk/ji2v3XENWFVYrf/K1jLmUmcwgA==",
            "type": "text",
          },
          "bounce": false,
          "from": "@treasure(client)",
          "to": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 65789n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{C41949DF000000000000000000000000000000000000000000000000000000009502F9004_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
            "to": "kQCEU_9CZOdW0nEmPIFBeKm5TwUgQeZSSoGV5F-1NEfHVOro",
            "type": "internal",
            "value": 923104000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 2,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{FFEB40DE000000000000000000000000000000000000000000000000000000009502F900400A128FE2F818AF4A1202DCB07C4374ED1EF1F35DC49E9FAD6BD48BDFB08E74C40C_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCEU_9CZOdW0nEmPIFBeKm5TwUgQeZSSoGV5F-1NEfHVOro",
          "to": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
          "type": "internal",
          "value": 810646000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 18124n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Success",
              "type": "text",
            },
            "bounce": false,
            "from": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
            "to": "@treasure(client)",
            "type": "internal",
            "value": 5791334000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`Payouts should withdraw 1`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "UCVAvkAKQCCn7CYC5oDzJww7/mYFJTKuGwzv7XbfuTQhsZXwBwjwd+ai9d8SUHujsjk/ji2v3XENWFVYrf/K1jLmUmcwgA==",
            "type": "text",
          },
          "bounce": false,
          "from": "@treasure(client)",
          "to": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 65561n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{C41949DF000000000000000000000000000000000000000000000000000000012A05F2004_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
            "to": "kQCEU_9CZOdW0nEmPIFBeKm5TwUgQeZSSoGV5F-1NEfHVOro",
            "type": "internal",
            "value": 923332000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 2,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{FFEB40DE000000000000000000000000000000000000000000000000000000012A05F200400A128FE2F818AF4A1202DCB07C4374ED1EF1F35DC49E9FAD6BD48BDFB08E74C40C_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCEU_9CZOdW0nEmPIFBeKm5TwUgQeZSSoGV5F-1NEfHVOro",
          "to": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
          "type": "internal",
          "value": 810874000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 18124n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Success",
              "type": "text",
            },
            "bounce": false,
            "from": "kQBksBno6LW4bBwcshk5mhsJL-rBtRhxc1x4m8gg4DJjlvpq",
            "to": "@treasure(client)",
            "type": "internal",
            "value": 10791562000n,
          },
        ],
      },
    ],
  },
]
`;
